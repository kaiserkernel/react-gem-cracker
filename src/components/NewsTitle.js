import React, { Component } from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import url from 'url';
import { gql } from 'react-apollo';

class NewsTitle extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    url: PropTypes.string,
    rank: PropTypes.number,
    isRankVisible: PropTypes.bool,
  }
  static defaultProps = {
    // text: undefined,
    url: undefined,
    rank: undefined,
    isRankVisible: true,
  }
  static fragments = {
    newsItem: gql`
      fragment NewsTitle on NewsItem {
        id,
        title,
        url,
        rank
      }
    `,
  };

  upvote() {
    console.log(this);
  }
  hidestory() {
    console.log(this);
  }
  render() {
    return (
      <tr className="athing" id={this.props.id}>
        <td style={{ textAlign: 'right', verticalAlign: 'top' }} className="title">
          <span className="rank">{this.props.isRankVisible && `${this.props.rank}.`}</span>
        </td>
        <td style={{ verticalAlign: 'top' }} className="votelinks">
          <center>
            <a onClick={this.upvote} href="vote?id=15077519&amp;how=up&amp;auth=b73e5ad6975f51978fed805f4c3c079e9516fe1d&amp; goto=news"><div className="votearrow" title="upvote" /></a>
          </center>
        </td>
        <td className="title">
          <a className="storylink" href={this.props.url ? this.props.url : `item?id=${this.props.id}`}>
            {this.props.title}
          </a>
          {
            this.props.url &&
            <span className="sitebit comhead"> (
              <a href={`from?site=${url.parse(this.props.url).hostname}`}>
                <span className="sitestr">{url.parse(this.props.url).hostname}</span>
              </a>)
            </span>
          }
        </td>
      </tr>
    );
  }
}

export default NewsTitle;
