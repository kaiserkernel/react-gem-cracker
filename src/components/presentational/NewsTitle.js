import React, { Component } from 'react';
import gql from 'graphql-tag';
import Link from 'next/link';
import PropTypes from 'prop-types';
import url from 'url';

export default class NewsTitle extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    isRankVisible: PropTypes.bool,
    isUpvoteVisible: PropTypes.bool,
    rank: PropTypes.number,
    title: PropTypes.string.isRequired,
    url: PropTypes.string,
    upvoted: PropTypes.bool.isRequired,
    upvoteNewsItem: PropTypes.func.isRequired,
  }
  static defaultProps = {
    isRankVisible: true,
    isUpvoteVisible: true,
    rank: undefined,
    url: undefined,
  }
  static fragments = {
    newsItem: gql`
      fragment NewsTitle on NewsItem {
        id
        title
        url
        upvoted
      }
    `,
  };
  render() {
    return (
      <tr className="athing" id={this.props.id}>
        <td style={{ textAlign: 'right', verticalAlign: 'top' }} className="title">
          <span className="rank">{this.props.isRankVisible && `${this.props.rank}.`}</span>
        </td>
        <td style={{ verticalAlign: 'top' }} className="votelinks">
          <center>
            {
              this.props.isUpvoteVisible &&
              <a className={this.props.upvoted ? 'nosee' : ' '} onClick={() => this.props.upvoteNewsItem(this.props.id)} href="javascript:void(0)" >
                <div className="votearrow" title="upvote" />
              </a>
            }
          </center>
        </td>
        <td className="title">
          <a className="storylink" href={this.props.url ? this.props.url : `item?id=${this.props.id}`}>
            {this.props.title}
          </a>
          {
            this.props.url &&
            <span className="sitebit comhead"> (
              <a href={`from?site=${url.parse(this.props.url).hostname}`}>
                <span className="sitestr">{url.parse(this.props.url).hostname}</span>
              </a>)
            </span>
          }
        </td>
      </tr>
    );
  }
}
