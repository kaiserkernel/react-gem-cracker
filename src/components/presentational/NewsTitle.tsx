import gql from 'graphql-tag';
import * as React from 'react';
import { parse } from 'url';

export class NewsTitle extends React.Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    isRankVisible: PropTypes.bool,
    isUpvoteVisible: PropTypes.bool,
    rank: PropTypes.number,
    title: PropTypes.string.isRequired,
    url: PropTypes.string,
    upvoted: PropTypes.bool.isRequired,
    upvoteNewsItem: PropTypes.func.isRequired,
  };
  static defaultProps = {
    isRankVisible: true,
    isUpvoteVisible: true,
    rank: undefined,
    url: undefined,
  };
  static fragments = {
    newsItem: gql`
      fragment NewsTitle on NewsItem {
        id
        title
        url
        upvoted
      }
    `,
  };
  render(): JSX.Element {
    return (
      <tr className="athing" id={this.props.id}>
        <td style={{ textAlign: 'right', verticalAlign: 'top' }} className="title">
          <span className="rank">{this.props.isRankVisible && `${this.props.rank}.`}</span>
        </td>
        <td style={{ verticalAlign: 'top' }} className="votelinks">
          <div style={{ textAlign: 'center' }}>
            {this.props.isUpvoteVisible && (
              <a
                className={this.props.upvoted ? 'nosee' : ' '}
                onClick={() => this.props.upvoteNewsItem(this.props.id)}
                href="javascript:void(0)"
              >
                <div className="votearrow" title="upvote" />
              </a>
            )}
          </div>
        </td>
        <td className="title">
          <a className="storylink" href={this.props.url ? this.props.url : `item?id=${this.props.id}`}>
            {this.props.title}
          </a>
          {this.props.url && (
            <span className="sitebit comhead">
              {' '}
              (
              <a href={`from?site=${parse(this.props.url).hostname}`}>
                <span className="sitestr">{parse(this.props.url).hostname}</span>
              </a>
              )
            </span>
          )}
        </td>
      </tr>
    );
  }
}
