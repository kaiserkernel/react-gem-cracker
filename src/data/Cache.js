// Interface: Commentable (Object can be commented on) comments, commentCount, commenter
// Interface: Voteable (Object can be voted on) upvotes, upvoteCount, hidden, hiddenCount,
// Data Type: Comment, can be on a news item or another comment
// Every time an upvote/downvote/comment is made, update the count

import LRU from 'lru-cache';

import {
  Feed,
} from './models';

// The cache is a singleton

export function warmCache() {
  // Fetch the front pages
  Feed.getForType('HOT', 30, 0);
  Feed.getForType('NEW', 30, 0);
  Feed.getForType('TOP', 30, 0);

  // Run every 15 mins
  setTimeout(warmCache, 1000 * 60 * 15);
}

class Cache {
  isReady = false;

  /*                  BEGIN NEWS ITEMS                      */


  getNewsItem(id) {
    // return this.newsItems.find(newsItem => newsItem.id === id);
    return this.newsItemsCache.get(id);
  }
  setNewsItem(newsItem) {
    return this.newsItemsCache.set(newsItem.id, newsItem);
  }

  /*                  END NEWS ITEMS                      */

  /*                   BEGIN USERS                        */


  getUser(id) {
    return this.userCache.get(id);
  }

  getUsers() {
    return this.userCache.dump();
  }

  createUser(user) {
    this.userCache.set(user.id, user);
    return user;
  }

  /*                    END USERS                         */

  /*                   BEGIN CACHES                         */


  newNewsItemsCache = LRU({
    max: 500,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  })
  newsItemsCache = LRU({
    max: 1000,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  })
  userCache = LRU({
    max: 500,
    maxAge: 1000 * 60 * 60 * 2, // 2 hour cache: ms * s * m
  })

  /*                   END CACHES                         */
}

export default new Cache();
