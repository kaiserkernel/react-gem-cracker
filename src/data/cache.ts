import { debug } from 'debug';
import LRU from 'lru-cache';

import { FeedSingleton, NewsItem } from './models';
import { FeedType } from './models/feed';
import { User } from './models/user';
import { Comment } from './models/comment';

const logger = debug('app:Cache');
logger.log = console.log.bind(console);

// The cache is a singleton

export function warmCache(): void {
  // Fetch the front pages
  FeedSingleton.getForType(FeedType.TOP, 30, 0);
  FeedSingleton.getForType(FeedType.NEW, 30, 0);

  // Run every 15 mins
  setTimeout(warmCache, 1000 * 60 * 15);
}

class Cache {
  isReady = false;

  /*                  BEGIN NEWS ITEMS                      */

  getNewsItem(id: number): NewsItem | undefined {
    return this.newsItemsCache.get(id.toString());
  }

  setNewsItem(id: number, newsItem: NewsItem): boolean {
    return this.newsItemsCache.set(id.toString(), newsItem);
  }

  /*                  END NEWS ITEMS                      */

  /*                   BEGIN USERS                        */

  getUser(id): User | undefined {
    return this.userCache.get(id);
  }

  getUsers(): Array<LRU.Entry<string, User>> {
    return this.userCache.dump();
  }

  setUser(id: string, user: User): User {
    logger(`Cache set user ${user}`);
    this.userCache.set(id, user);
    return user;
  }

  /*                    END USERS                         */

  /*                   BEGIN COMMENTS                        */

  getComment(id: number): Comment | undefined {
    return this.commentCache.get(id.toString());
  }

  setComment(id: number, comment: Comment): Comment {
    this.commentCache.set(comment.id.toString(), comment);
    logger(`Cache set comment ${comment}`);
    return comment;
  }

  /*                    END COMMENTS                         */

  /*                   BEGIN CACHES                         */

  newNewsItemsCache = new LRU<string, NewsItem>({
    max: 500,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  });

  newsItemsCache = new LRU<string, NewsItem>({
    max: 1000,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  });

  userCache = new LRU<string, User>({
    max: 500,
    maxAge: 1000 * 60 * 60 * 2, // 2 hour cache: ms * s * m
  });

  commentCache = new LRU<string, Comment>({
    max: 5000,
    maxAge: 1000 * 60 * 60 * 1, // 1 hour cache: ms * s * m
  });

  /*                   END CACHES                         */
}

export const cache = new Cache();
