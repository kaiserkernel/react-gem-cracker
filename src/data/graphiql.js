/**  
   * WARNING: THIS USES THE REAL SCHEMA AND DATABASE!
   *
   *   Take care not to cause data loss but the app should be secure
   */

// TODO: Can make this graphql server part of the main server using feature flag
// Example GraphiQL server https://github.com/graphql/graphiql/tree/master/example

import express from 'express';
import bodyParser from 'body-parser';
import {
  graphqlExpress,
  graphiqlExpress,
} from 'apollo-server-express';

import Schema from './Schema';
import {
  APP_PORT,
  graphQLPath,
  graphiQLPath,
  GRAPHQL_URL,
  GRAPHIQL_URL,
} from '../config';
import {
  Comment,
  Feed,
  NewsItem,
  User,
} from './models';


const logger = console; // global.logger;

const app = express();
// app.use(morgan('dev'));

// Mount GraphQL Server middleware
app.use(graphQLPath, bodyParser.json(), graphqlExpress(request => ({
  schema: Schema,
  rootValue: { request },
  context: {
    Comment,
    Feed,
    NewsItem,
    User,
  },
  debug: true,
})));

app.use(graphiQLPath, graphiqlExpress({
  endpointURL: graphQLPath,
}));

app.listen(APP_PORT, () => {
  logger.info(`==> ðŸ“ˆ GraphQL Server on ${GRAPHQL_URL}`);
  logger.info(`==> ðŸŒŽ  Go to ${GRAPHIQL_URL}`);
});
