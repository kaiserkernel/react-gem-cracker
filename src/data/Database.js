import data from './SampleData';
import cache from './Cache';

/*                  BEGIN NEWS ITEMS                      */


export function getNewsItem(id) {
  return data.newsItems.find(newsItem => newsItem.id === id);
}

export function rankNewsItems() {
  // Would implement this somewhere in the real HN system
  // A scheduled job to recalculate post rank and points
  // .sort((a, b) => (a.rank - b.rank))
}

export function createNewsItem(newsItem) {
  data.newsItems.push(newsItem);
  return newsItem;
}

//                  NEWS ITEM MUTATIONS


export function upvoteNewsItem(id, userId) {
  const newsItemData = data.newsItems.find(newsItem => newsItem.id === id);
  newsItemData.upvotes.add(userId);
  newsItemData.upvoteCount += 1;

  return newsItemData;
}

export function downvoteNewsItem(id, userId) {
  const newsItemData = data.newsItems.find(newsItem => newsItem.id === id);
  newsItemData.downvotes.add(userId);
  newsItemData.downvoteCount += 1;

  return newsItemData;
}

/*                  END NEWS ITEMS                      */

/*                     BEGIN FEED                         */


export function getNewNewsItems(first, skip) {
  return data.new.slice(skip, skip + first)
    .map((postId, index) => ({
      ...getNewsItem(postId),
      rank: skip + index + 1,
    }));
}

export function getTopNewsItems(first, skip) {
  return data.newsItems.slice(skip, skip + first);
}

export function getHotNews() {
  return data.newsItems;
}

export function getNewsItems() {
  return data.newsItems;
}

/*                     END FEED                         */

/*                   BEGIN USERS                        */


export function getUser(id) {
  return data.users.find(user => user.id === id);
}

export function getUsers() {
  return data.users;
}

export function createUser(user) {
  data.users.push(user);
  return user;
}

/*                    END USERS                         */
