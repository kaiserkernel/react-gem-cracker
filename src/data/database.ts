import { debug } from 'debug';

import { cache } from './cache';
import { FeedSingleton, User } from './models';
import { sampleData } from './sample-data';
import { NewsItem } from './models/news-item';

const logger = debug('app:Database');
logger.log = console.log.bind(console);

/*                  BEGIN NEWS ITEMS                      */

export function getNewsItem(id: number): NewsItem | undefined {
  return sampleData.newsItems.find((newsItem) => newsItem.id === id);
}

export function createNewsItem(newsItem: NewsItem): NewsItem {
  sampleData.newsItems.push(newsItem);

  return newsItem;
}

//                  NEWS ITEM MUTATIONS

export function upvoteNewsItem(id: number, userId: string): NewsItem | undefined {
  // Upvote the News Item in the DB
  const newsItem = cache.getNewsItem(id);

  if (newsItem && !newsItem.upvotes.includes(userId)) {
    newsItem.upvotes.push(userId);
    newsItem.upvoteCount += 1;
    cache.setNewsItem(id, newsItem);
  }

  return newsItem;
}

export function unvoteNewsItem(id: number, userId: string): NewsItem | undefined {
  const newsItem = cache.getNewsItem(id);

  if (newsItem && !newsItem.upvotes.includes(userId)) {
    newsItem.upvotes.splice(newsItem.upvotes.indexOf(userId), 1);
    newsItem.upvoteCount -= 1;
    cache.setNewsItem(id, newsItem);
  }

  return newsItem;
}

export function hideNewsItem(id: number, userId: string): NewsItem {
  logger(`Hiding News Item ${id} by ${userId}`);

  const newsItem = cache.getNewsItem(id);
  const user = cache.getUser(userId);

  if (user && !user.hides.includes(id) && newsItem && !newsItem.hides.includes(userId)) {
    user.hides.push(id);
    cache.setUser(userId, user);

    newsItem.hides.push(userId);
    cache.setNewsItem(id, newsItem);

    logger(`Hid News Item ${id} by ${userId}`);
  } else {
    throw new Error(`Data error, user has already hidden ${id} by ${userId}`);
  }

  return newsItem;
}

export function submitNewsItem(id: number, newsItem: NewsItem) {
  // Submit the News Item in the DB
  if (cache.setNewsItem(id, newsItem)) {
    FeedSingleton.new.unshift(id);
    FeedSingleton.new.pop();
    return newsItem;
  }

  throw new Error('Unable to submit News Item.');
}

/*                  END NEWS ITEMS                      */

/*                     BEGIN FEED                         */

export function getNewNewsItems(first: number, skip: number) {
  return sampleData.new.slice(skip, skip + first).map((postId, index) => ({
    ...getNewsItem(postId),
    rank: skip + index + 1,
  }));
}

export function getTopNewsItems(first: number, skip: number) {
  return sampleData.newsItems.slice(skip, skip + first);
}

export function getHotNews(): NewsItem[] {
  return sampleData.newsItems;
}

export function getNewsItems(): NewsItem[] {
  return sampleData.newsItems;
}

/*                     END FEED                         */

/*                   BEGIN USERS                        */

export function getUser(id: string) {
  return sampleData.users.find((user) => user.id === id);
}

export function getUsers() {
  return sampleData.users;
}

export function createUser(user: User) {
  sampleData.users.push(user);

  return user;
}

/*                    END USERS                         */
