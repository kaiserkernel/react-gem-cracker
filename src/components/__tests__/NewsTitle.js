import React from 'react';
// import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import NewsTitle from '../NewsTitle';
import data from '../../data/SampleData';


describe('NewsTitle component', () => {
  // it('shallow renders', () => {
  //   const wrapper = shallow(
  //     <NewsFeed />,
  //   );
  //   expect(wrapper).toMatchSnapshot();
  // });
  it('renders news item properties passed in as props', () => {
    const wrapper = shallow((
      <NewsTitle {...data.newsItems[0]} rankVisible={true} />
    ));
    expect(wrapper).toMatchSnapshot();
  });
  // it('renders news items passed in as props', () => {
  //   const wrapper = shallow((
  //     <NewsFeed newsItems={data.newsItems} />
  //   ));
  //   expect(wrapper.contains(<div className="unique" />)).to.equal(true);
  // });
});
