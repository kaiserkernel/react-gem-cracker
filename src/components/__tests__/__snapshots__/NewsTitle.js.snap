// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsTitle component renders news item properties passed in as props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <tr
    className="athing"
    id={1224}
>
    <td
        className="title"
        style={
            Object {
                "textAlign": "right",
                "verticalAlign": "top",
              }
        }
    >
        <span
            className="rank"
        >
            1.
        </span>
    </td>
    <td
        className="votelinks"
        style={
            Object {
                "verticalAlign": "top",
              }
        }
    >
        <center>
            <a
                href="vote?id=15077519&how=up&auth=b73e5ad6975f51978fed805f4c3c079e9516fe1d& goto=news"
                onClick={[Function]}
            >
                <div
                    className="votearrow"
                    title="upvote"
                />
            </a>
        </center>
    </td>
    <td
        className="title"
    >
        <a
            className="storylink"
            href="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
        >
            “Learning How to Learn,” the most popular course on Coursera
        </a>
        <span
            className="sitebit comhead"
        >
             (
            <a
                href="from?site=www.nytimes.com"
            >
                <span
                    className="sitestr"
                >
                    www.nytimes.com
                </span>
            </a>
            )
        </span>
    </td>
</tr>,
  "nodes": Array [
    <tr
      className="athing"
      id={1224}
>
      <td
            className="title"
            style={
                  Object {
                        "textAlign": "right",
                        "verticalAlign": "top",
                      }
            }
      >
            <span
                  className="rank"
            >
                  1.
            </span>
      </td>
      <td
            className="votelinks"
            style={
                  Object {
                        "verticalAlign": "top",
                      }
            }
      >
            <center>
                  <a
                        href="vote?id=15077519&how=up&auth=b73e5ad6975f51978fed805f4c3c079e9516fe1d& goto=news"
                        onClick={[Function]}
                  >
                        <div
                              className="votearrow"
                              title="upvote"
                        />
                  </a>
            </center>
      </td>
      <td
            className="title"
      >
            <a
                  className="storylink"
                  href="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
            >
                  “Learning How to Learn,” the most popular course on Coursera
            </a>
            <span
                  className="sitebit comhead"
            >
                   (
                  <a
                        href="from?site=www.nytimes.com"
                  >
                        <span
                              className="sitestr"
                        >
                              www.nytimes.com
                        </span>
                  </a>
                  )
            </span>
      </td>
</tr>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <NewsTitle
        commentCount={44}
        comments={Array []}
        creationTime={2017-09-21T20:12:16.000Z}
        hidden={Array []}
        hiddenCount={0}
        id={1224}
        points={171}
        rank={1}
        rankVisible={true}
        submitterId="hvo"
        text={null}
        title="“Learning How to Learn,” the most popular course on Coursera"
        upvoteCount={10}
        upvotes={Set {}}
        url="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": NewsTitle {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "commentCount": 44,
          "comments": Array [],
          "creationTime": 2017-09-21T20:12:16.000Z,
          "hidden": Array [],
          "hiddenCount": 0,
          "id": 1224,
          "points": 171,
          "rank": 1,
          "rankVisible": true,
          "submitterId": "hvo",
          "text": null,
          "title": "“Learning How to Learn,” the most popular course on Coursera",
          "upvoteCount": 10,
          "upvotes": Set {},
          "url": "https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <tr
          className="athing"
          id={1224}
>
          <td
                    className="title"
                    style={
                              Object {
                                        "textAlign": "right",
                                        "verticalAlign": "top",
                                      }
                    }
          >
                    <span
                              className="rank"
                    >
                              1.
                    </span>
          </td>
          <td
                    className="votelinks"
                    style={
                              Object {
                                        "verticalAlign": "top",
                                      }
                    }
          >
                    <center>
                              <a
                                        href="vote?id=15077519&how=up&auth=b73e5ad6975f51978fed805f4c3c079e9516fe1d& goto=news"
                                        onClick={[Function]}
                              >
                                        <div
                                                  className="votearrow"
                                                  title="upvote"
                                        />
                              </a>
                    </center>
          </td>
          <td
                    className="title"
          >
                    <a
                              className="storylink"
                              href="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
                    >
                              “Learning How to Learn,” the most popular course on Coursera
                    </a>
                    <span
                              className="sitebit comhead"
                    >
                               (
                              <a
                                        href="from?site=www.nytimes.com"
                              >
                                        <span
                                                  className="sitestr"
                                        >
                                                  www.nytimes.com
                                        </span>
                              </a>
                              )
                    </span>
          </td>
</tr>,
        "_debugID": 2,
        "_renderedOutput": <tr
          className="athing"
          id={1224}
>
          <td
                    className="title"
                    style={
                              Object {
                                        "textAlign": "right",
                                        "verticalAlign": "top",
                                      }
                    }
          >
                    <span
                              className="rank"
                    >
                              1.
                    </span>
          </td>
          <td
                    className="votelinks"
                    style={
                              Object {
                                        "verticalAlign": "top",
                                      }
                    }
          >
                    <center>
                              <a
                                        href="vote?id=15077519&how=up&auth=b73e5ad6975f51978fed805f4c3c079e9516fe1d& goto=news"
                                        onClick={[Function]}
                              >
                                        <div
                                                  className="votearrow"
                                                  title="upvote"
                                        />
                              </a>
                    </center>
          </td>
          <td
                    className="title"
          >
                    <a
                              className="storylink"
                              href="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
                    >
                              “Learning How to Learn,” the most popular course on Coursera
                    </a>
                    <span
                              className="sitebit comhead"
                    >
                               (
                              <a
                                        href="from?site=www.nytimes.com"
                              >
                                        <span
                                                  className="sitestr"
                                        >
                                                  www.nytimes.com
                                        </span>
                              </a>
                              )
                    </span>
          </td>
</tr>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <NewsTitle
    commentCount={44}
    comments={Array []}
    creationTime={2017-09-21T20:12:16.000Z}
    hidden={Array []}
    hiddenCount={0}
    id={1224}
    points={171}
    rank={1}
    rankVisible={true}
    submitterId="hvo"
    text={null}
    title="“Learning How to Learn,” the most popular course on Coursera"
    upvoteCount={10}
    upvotes={Set {}}
    url="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
/>,
}
`;
